### PR TITLE
Feature: Repository name filter

### PR DESCRIPTION
**Pull request by butzist**
**Purpose of the PR**

- Allow specifying a regex for matching multiple/all repos in a filter rule

Fixes #

- https://github.com/Azure/acr-cli/issues/57

### CODE REVIEW

File: cmd/acr/purge_test.go [hunk 1]
```diff
@@ -559,6 +559,70 @@ func TestDryRun(t *testing.T) {
...
+		assert.Equal(".*", filters[testRepo], "Filter for test repo should be .*")
+		assert.Equal(".*", filters["bar"], "Filter for bar repo should be .*")
+		assert.Equal(nil, err, "Error should be nil")
+		mockClient.AssertExpectations(t)
+	})
+	t.Run("AllReposWildcard", func(t *testing.T) {
+		assert := assert.New(t)
+		mockClient := &mocks.BaseClientAPI{}
+		mockClient.On("GetRepositories", mock.Anything, "", (*int32)(nil)).Return(ManyRepositoriesResult, nil).Once()
+		filters, err := collectTagFilters(testCtx, []string{":.*"}, mockClient)
```
Comment (Wwwsylvia) [inline cmd/acr/purge_test.go:590]: What's the difference between this case and the last one? Do you mean `".*:.*"` here?

File: cmd/acr/purge.go [hunk 2]
```diff
@@ -211,6 +200,56 @@ func purgeTags(ctx context.Context, acrClient api.AcrCLIClientInterface, poolSiz
...
+
+	tagFilters := map[string]string{}
+	for _, filter := range rawFilters {
+		repoRegex, tagRegex, err := getRepositoryAndTagRegex(filter)
+		if err != nil {
+			return nil, err
+		}
+		var repoNames []string
+		if repoRegex == "" {
+			repoNames = *allRepoNames.Names
```
Comment (Wwwsylvia) [inline cmd/acr/purge.go:218]: This looks dangerous... People may purge all the repositories/tags by accidentally passing `":"`, `"repo:"`, or `":tags"`. Should we not allow empty repoRegex and empty tagRegex? Is it originally intended? /cc @gildardogmsft

File: cmd/acr/purge.go [hunk 3]
```diff
@@ -28,6 +29,9 @@ const (
 	purgeExampleMessage    = `  - Delete all tags that are older than 1 day in the example.azurecr.io registry inside the hello-world repository
     	acr purge -r example --filter "hello-world:.*" --ago 1d
 
+  - Delete all tags that are older than 7 days in the example.azurecr.io registry inside all repositories
+	    acr purge -r example --filter ".*:.*" --ago 7d
```
Comment (Wwwsylvia) [inline cmd/acr/purge.go:33]: Can we also update the documentation in README? 🙂

**Comment by Wwwsylvia** [comment]
> What's the issue with the build? Anything on my side?

Thanks @butzist for contributing to this repo!  The build was broken due to https://github.com/Azure/acr-cli/issues/89 and we've fixed it. You may want to rebase if it still not working.

File: cmd/acr/purge.go (hunk 2 above)
Comment (butzist) [reply cmd/acr/purge.go:218]: This variant was related to the following comment: https://github.com/Azure/acr-cli/issues/57#issuecomment-777899936
I found this way for specifying all repos pretty convenient :)

File: cmd/acr/purge_test.go (hunk 1 above)
Comment (butzist) [reply cmd/acr/purge_test.go:590]: Related to your comment below:
https://github.com/Azure/acr-cli/pull/88#discussion_r804442611

Please decide if this is desirable or not

File: cmd/acr/purge.go (hunk 2 above)
Comment (gildardogmsft) [reply cmd/acr/purge.go:218]: Hmmm I agree with @Wwwsylvia having an empty regex meaning all repositories sounds a little bit dangerous to me, I would prefer to not allow empty repoRegex and also not allow empty tagRegex

File: cmd/acr/purge_test.go (hunk 1 above)
Comment (butzist) [reply cmd/acr/purge_test.go]: Fixed

File: cmd/acr/purge.go (hunk 2 above)
Comment (butzist) [reply cmd/acr/purge.go]: Added errors on missing tag or repo regex

File: cmd/acr/purge.go (hunk 3 above)
Comment (butzist) [reply cmd/acr/purge.go:33]: Done, please check

**Comment by butzist** [comment]
@Wwwsylvia @gildardogmsft Good morning, it's ready for re-review.

File: cmd/acr/purge.go (hunk 2 above)
Comment (butzist) [reply cmd/acr/purge.go]: image

**Comment by Wwwsylvia** [comment]
Thanks @butzist.
Another thing I noticed is that the tag regex allows partial match (@gildardogmsft Is this by design?), while the repo regex does not. Would this confuse users? Should we keep them consistent? I'm not very sure which way users prefer though.

**Comment by butzist** [comment]
@Wwwsylvia The repo regex needs to match the complete name, otherwise we might match too many repos by accident, and mostly preserve backwards compatibility. I left the tag to be partially matching, for the same reason - backwards compatibility. If you want to break compatibility, please do this in a separate PR.

**Comment by Wwwsylvia** [comment]
Thanks @butzist for the clarification! That makes sense.

**Comment by butzist** [comment]
Thanks for the review, @Wwwsylvia

**Pull request merged**